### PR TITLE
Blockfamily library handler

### PR DESCRIPTION
**Pull request by pollend**
I replaced BlockFamilyRegistry with a block family library type. The idea behind this change is to allow a block family to be mapped to the module vs just the name. This is more specific and should avoid problems in the future where a family is named the same in two separate modules. There are currently some block families that follow the module:family type. This can be tested with some of the blocks under the additional rails module where some of the special rail blocks use the family type from the base rails module.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

File: engine/src/main/java/org/terasology/math/Vector3fUtil.java [hunk 1]
```diff
@@ -65,18 +65,5 @@ public static Vector3f safeNormalize(Vector3f v, Vector3f out) {
...
-        out.z = Math.min(a.z, b.z);
-        return out;
-    }
-
-    public static Vector3f max(Vector3f a, Vector3f b, Vector3f out) {
-        out.x = Math.max(a.x, b.x);
-        out.y = Math.max(a.y, b.y);
-        out.z = Math.max(a.z, b.z);
-        return out;
-    }
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/math/Vector3fUtil.java]: This also shouldn't be here?

File: engine/src/main/java/org/terasology/world/block/loader/BlockFamilyDefinition.java [hunk 2]
```diff
@@ -48,17 +48,17 @@ protected void doReload(BlockFamilyDefinitionData blockFamilyDefinitionData) {
...
 
     public boolean isFreeform() {
-        return BlockFamilyRegistry.isFreeformSupported(getData().getBlockFamily());
+        return BlockfamilyLibrary.isFreeformSupported(getData().getBlockFamily());
     }
 
     public BlockFamily createFamily(BlockBuilderHelper blockBuilderHelper) {
         Preconditions.checkState(!isFreeform());
-        return BlockFamilyRegistry.createFamily(getData().getBlockFamily(), this, blockBuilderHelper);
+        return BlockfamilyLibrary.createFamily(getData().getBlockFamily(), this, blockBuilderHelper);
```
Comment (jellysnake) [inline engine/src/main/java/org/terasology/world/block/loader/BlockFamilyDefinition.java]: Why the lowercase `family` change?

**Comment by jellysnake** [comment]
Also appears you missed a spot in the tests with the changes:

```
> Task :engine-tests:compileJava
/home/jenkins/workspace/TerasologyPRs/engine-tests/src/main/java/org/terasology/HeadlessEnvironment.java:208: error: constructor BlockFamilyDefinitionFormat in class BlockFamilyDefinitionFormat cannot be applied to given types;
                new BlockFamilyDefinitionFormat(assetTypeManager.getAssetManager(), context.get(BlockfamilyLibrary.class)));
                ^
  required: AssetManager
  found: AssetManager,BlockfamilyLibrary
  reason: actual and formal argument lists differ in length
```

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by pollend** [comment]
#3418

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Tested this out a bit. Does seem to make the previously misbehaving rail blocks in AdditionalRails function. Also throws occasional NPEs (benign?) when something fails to deserialize somehow:

```
21:45:46.907 [Chunk-Generator-7] INFO  o.t.w.b.internal.BlockManagerImpl - Registered Block core:Cactus with id 56
21:45:46.908 [Chunk-Generator-7] INFO  o.t.w.b.internal.BlockManagerImpl - Registered BlockFamily[freeform,core:PineTrunk]
21:45:46.909 [Chunk-Generator-7] INFO  o.t.w.b.internal.BlockManagerImpl - Registered Block core:PineTrunk with id 57
21:45:48.026 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
	at org.terasology.world.block.family.BlockFamilyLibrary.getBlockFamily(BlockFamilyLibrary.java:71)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyHandler.deserialize(BlockFamilyDefinitionFormat.java:335)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyHandler.deserialize(BlockFamilyDefinitionFormat.java:327)
	at com.google.gson.TreeTypeAdapter.read(TreeTypeAdapter.java:58)
	at com.google.gson.Gson.fromJson(Gson.java:879)
	at com.google.gson.Gson.fromJson(Gson.java:944)
	at com.google.gson.Gson$1.deserialize(Gson.java:138)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.setObject(BlockFamilyDefinitionFormat.java:303)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.deserialize(BlockFamilyDefinitionFormat.java:143)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.deserialize(BlockFamilyDefinitionFormat.java:129)
	at com.google.gson.TreeTypeAdapter.read(TreeTypeAdapter.java:58)
	at com.google.gson.Gson.fromJson(Gson.java:879)
	at com.google.gson.Gson.fromJson(Gson.java:817)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat.load(BlockFamilyDefinitionFormat.java:94)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat.load(BlockFamilyDefinitionFormat.java:71)
	at org.terasology.assets.module.UnloadedAssetData$AssetSourceResolver.load(UnloadedAssetData.java:292)
	at org.terasology.assets.module.UnloadedAssetData.load(UnloadedAssetData.java:180)
	at org.terasology.assets.module.ModuleAssetDataProducer.getAssetData(ModuleAssetDataProducer.java:228)
	at org.terasology.assets.AssetType.lambda$reload$2(AssetType.java:359)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.terasology.assets.AssetType.reload(AssetType.java:357)
	at org.terasology.assets.AssetType.getNormalAsset(AssetType.java:386)
	at org.terasology.assets.AssetType.getAsset(AssetType.java:262)
	at org.terasology.assets.management.AssetManager.getAsset(AssetManager.java:223)
	at org.terasology.world.block.internal.BlockManagerImpl.loadFamily(BlockManagerImpl.java:270)
	at org.terasology.world.block.internal.BlockManagerImpl.getBlockFamily(BlockManagerImpl.java:247)
	at org.terasology.world.block.internal.BlockManagerImpl.getBlockFamily(BlockManagerImpl.java:232)
	at org.terasology.gookeeper.system.PlayerStartingInventorySystem.onPlayerSpawnedEvent(PlayerStartingInventorySystem.java:47)
	at org.terasology.gookeeper.system.PlayerStartingInventorySystemMethodAccess.invoke(Unknown Source)
	at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:528)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:288)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:279)
	at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:204)
	at org.terasology.logic.players.PlayerSystem.spawnPlayer(PlayerSystem.java:274)
	at org.terasology.logic.players.PlayerSystem.update(PlayerSystem.java:87)
	at org.terasology.engine.modes.loadProcesses.AwaitCharacterSpawn.step(AwaitCharacterSpawn.java:45)
	at org.terasology.engine.modes.StateLoading.update(StateLoading.java:238)
	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:450)
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:413)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:389)
	at org.terasology.engine.Terasology.main(Terasology.java:154)
21:45:48.978 [Chunk-Generator-2] INFO  o.t.w.b.internal.BlockManagerImpl - Registered BlockFamily[freeform,core:RedLeaf]
21:45:48.978 [Chunk-Generator-2] INFO  o.t.w.b.internal.BlockManagerImpl - Registered Block core:RedLeaf with id 58
```

Everything still compiles in a mega workspace and there are no usages I can find for `BlockFamilyRegistry` - which is kinda odd since it does appear to be API. Looks like it is only used within the engine.

There's also a weird `target/.history` included which should be removed + added to the `.gitignore` :-)

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Pull request merged**